// @hmcts/properties-volume/lib/LocalAzureVault.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addFromAzureVault = addFromAzureVault;
const identity_1 = require("@azure/identity");
const keyvault_secrets_1 = require("@azure/keyvault-secrets");
const yaml = require("js-yaml");
const nodejs_logging_1 = require("@hmcts/nodejs-logging");
const fs = require("fs");
const { merge } = require('lodash');
const log = nodejs_logging_1.Logger.getLogger('applicationRunner');
async function addFromAzureVault(config, options) {
    const env = options.env || 'aat';
    log.info(`Attempting to read properties from volume: '${options.pathToHelmChart}' using env: '${env}'`);
    if (!options.pathToHelmChart || !fs.existsSync(options.pathToHelmChart)) {
        throw new Error(`helm chart not found at: '${options.pathToHelmChart}'`);
    }
    try {
        const chart = yaml.load(fs.readFileSync(options.pathToHelmChart, 'utf8'));
        const secrets = await readVaultsFromAzure(chart, env);
        config['secrets'] = merge(config['secrets'] || {}, secrets);
        return config;
    }
    catch (error) {
        log.error(error);
        throw Error(`properties-volume failed with: ${error}`);
    }
}
async function readVaultsFromAzure(chart, env) {
    const credential = new identity_1.DefaultAzureCredential();
    const vaultSecrets = deepSearch(chart, 'keyVaults');
    if (!vaultSecrets) {
        throw new Error('No keyVaults found in helm chart');
    }
    const vaultPromises = Object.keys(vaultSecrets).map(vaultName => readVaultFromAzure(vaultSecrets[vaultName], vaultName, env, credential));
    const vaults = await Promise.all(vaultPromises);
    return merge({}, ...vaults);
}
async function readVaultFromAzure(vaultSecrets, vaultName, env, credential) {
    const vaultUri = `https://${vaultName}-${env}.vault.azure.net`;
    const client = new keyvault_secrets_1.SecretClient(vaultUri, credential);
    const chartSecrets = (vaultSecrets === null || vaultSecrets === void 0 ? void 0 : vaultSecrets.secrets) || [];
    const secretPromises = chartSecrets.map(secret => normalizeSecret(secret)).map(secret => loadSecret(client, secret));
    const loadedSecrets = await Promise.all(secretPromises);
    return { [vaultName]: merge({}, ...loadedSecrets) };
}
function deepSearch(obj, key) {
    if (obj.hasOwnProperty(key)) {
        return obj[key];
    }
    for (const k in obj) {
        if (typeof obj[k] === 'object') {
            const result = deepSearch(obj[k], key);
            if (result) {
                return result;
            }
        }
    }
}
function normalizeSecret(secret) {
    return {
        alias: (secret === null || secret === void 0 ? void 0 : secret.alias) || secret,
        name: (secret === null || secret === void 0 ? void 0 : secret.name) || secret,
    };
}
async function loadSecret(client, secret) {
    const secretValue = await client.getSecret(secret.name);
    return { [secret.alias]: secretValue.value };
}
